### PR TITLE
Language Server POC

### PR DESCRIPTION
**Pull request by dantleech**
This PR is a POC for a built-in Language Server for PHPStan  (as hinted at in [this ticket](https://github.com/phpstan/phpstan/issues/3262)) which supports **diagnostics**

I've been somewhat busy this week refactoring the Phpactor Language Server Platform and thought it would be interesting to try and integrate it with PHPStan.

Currently this is a naive implementation which works on document _save_ but has issues with the PHPStan (internal) cache.

Any final implementation should:

- Lint in real-time.
- Dispatching a (non-blocking and stateless) worker.
- Lints should be de-duplicated and queued.

See [phpactor-phpstan](https://github.com/phpactor/language-server-phpstan-extension/blob/master/lib/Handler/PhpstanHandler.php) for an approach which uses a sub-process (and runs as a background service)

Anyway, just playing about - let me know if this is interesting and/or feel free to close. I'd create a new PR anyway (this one is from `dantleech:master` :/ )

### CODE REVIEW

File: src/Command/LanguageServerCommand.php [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+                $input,
+                $output,
+                ['.'],
+                null,
+                null,
+                $autoloadFile,
+                $this->composerAutoloaderProjectPaths,
+                $configuration,
+                null,
+                '7',
```
Comment (dantleech) [inline src/Command/LanguageServerCommand.php:65]: hard coded this for now

File: src/Command/LanguageServerCommand.php [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        $container = $inceptionResult->getContainer();
+        assert($output instanceof ConsoleOutput);
+
+        $logger = new class extends AbstractLogger {
+            public function log($level, $message, array $context = array()) {
+                fwrite(STDERR, sprintf(
+                    '[%s] %s ::: %s', $level, $message, json_encode($context, JSON_PRETTY_PRINT)
+                ) . "\n");
+            }
+        };
```
Comment (dantleech) [inline src/Command/LanguageServerCommand.php:82]: Could use the `ConsoleLogger` but it didn't log the `$context` would be replaced...

File: src/LanguageServer/DiagnosticsHandler.php [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    private function buildDiagnostics(AnalysisResult $result): array
+    {
+        $diagnostics = [];
+        var_dump($result);
+        foreach ($result->getFileSpecificErrors() as $error) {
+
+            $diagnostics[] = new Diagnostic(
+                new Range(
+                    new Position($error->getLine() - 1, 0),
+                    new Position($error->getLine() - 1, 100)
```
Comment (dantleech) [inline src/LanguageServer/DiagnosticsHandler.php:100]: As we don't have char positions in Phpstan (?) we would need to find out the start + end char of the line and highlight it all.

File: src/LanguageServer/DiagnosticsHandler.php [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    {
+        return [
+            DidSaveTextDocumentNotification::METHOD => 'didSave',
+        ];
+    }
+
+    /**
+     * @return Promise<null>
+     */
+    public function didSave(DidSaveTextDocumentParams $didSave): Promise
```
Comment (dantleech) [inline src/LanguageServer/DiagnosticsHandler.php:55]: We hijack the text-document sync notifications for our purposes - which is probably fine, but we can also use the platform's `TextDocumentHandler` and use the events it submits.

File: composer.json [hunk 5]
```diff
@@ -22,6 +22,7 @@
 		"nikic/php-parser": "dev-master",
 		"ondram/ci-detector": "^3.4.0",
 		"ondrejmirtes/better-reflection": "4.3.18",
+		"phpactor/language-server": "dev-lsp-prot-1",
```
Comment (dantleech) [inline composer.json:25]: this "new" version of the package hasn't been merged yet

File: src/LanguageServer/PhpstanDispatcherFactory.php [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+			new ErrorHandlingMiddleware($this->logger),
+			new InitializeMiddleware($handlers, new NullEventDispatcher()),
+			new HandlerMiddleware(
+				new HandlerMethodRunner(
+					$handlers,
+					new HandlerMethodResolver(),
+					$argumentResolver,
+				)
+			)
+		);
```
Comment (dantleech) [inline src/LanguageServer/PhpstanDispatcherFactory.php:64]: this class is responsible for creating a new "session". The server can be run either on a socket (allowing multiple connections - mostly useful for debugging) or over STDIO (which is how language servers are usually used in practice).

File: src/LanguageServer/DiagnosticsHandler.php [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    public function registerCapabiltiies(ServerCapabilities $capabilities): void
+    {
+        $syncOptions = new TextDocumentSyncOptions();
+        $syncOptions->save = new SaveOptions(true);
+        $syncOptions->willSave = true;
+        $syncOptions->change = TextDocumentSyncKind::FULL;
+        $capabilities->textDocumentSync = $syncOptions;
+    }
```
Comment (dantleech) [inline src/LanguageServer/DiagnosticsHandler.php:86]: the handler can set server capabilities which are sent back to the client as a response to it's initialization. I know what `change` does, not sure about the others honestly.

**Comment by ondrejmirtes** [comment]
Hi, I appreciate the effort, but I'd like it to be implemented a bit differently. Mainly, I'd need it to be build on top of `react/socket` instead of `amphp` (as that's my preferred async library and I already use it in PHPStan), and also I need it to support PHP 7.1 as well :)

But I'd like to play with this PR as it is to get the idea of what it can do, can you tell me the instructions what I need to do to try it out for example in VS Code? It's not obvious to me to how I should even run the `language-server` command, what's the difference between TCP address and stdio etc.

**Comment by dantleech** [comment]
I'll need to update it to the latest version of the LS library. VS code (ironically) is a pain because it's the only editor that requiess a typetypescript plugin per language server. Sublime is quite easy, I think Phpstorm has an LS plugin too.

But i'll try and update later, just about to start cycling up to Scotland... :)

**Comment by dantleech** [comment]
I've updated to the latest API changes.

- TCP mode basically runs a server that can handle multiple connections/sessions in the same process.
- STDIO means the client will start the process and send messages over STDIN and the server responds on STDOUT - it therefore only handles one session in the process.

In practice clients usually use the STDIO method.

For testing, you can use either method, but the TCP mode is a bit more transparent:

```
php7.4 ./bin/phpstan language-server --address=127.0.0.1:8890 
```

I'm using VIM with [CoC](https://github.com/neoclide/coc.nvim) the configuration is:

```
{
    // ...
    "languageserver": {
        "phpstan": {
            "enable": true,
            "revealOutputChannelOn": "never",
            "args": ["language-server"],
            "initializationOptions": {
            },
            "filetypes": ["php"],
            "port": 8890,
            "host": "0.0.0.0"
        },
        "psalmls": {
            "enable": false,
            "command": "vendor/bin/psalm-language-server",
            "args": [
                "--enable-autocomplete=false"
            ],
            "filetypes": ["php"],
            "rootPatterns": ["psalm.xml", "psalm.xml.dist"],
            "requireRootPattern": true
        }
    }
}
```
The `phpstan` config will attempt to connect on `0.0.0.0:8090` (`psalmls` is using STDIO above, left for example)

The config is essentially similar for other editors, check the documentation for [Sublime](https://lsp.readthedocs.io/en/stable/). I've heard there is an LSP plugin for IntelliJ. VSCode requires a custom plugin (!), I wrote [this one](https://github.com/phpactor/vscode-phpactor/blob/master/src/extension.ts) for Phpactor - maybe there's a generic plugin but I haven't found it.

Let me know if you want to have a hangout or similar at some point and I could explain it better.

**Comment by ondrejmirtes** [comment]
Hi, thank you for this effort, but the current approach isn't viable:

1) I'd need `phpactor/language-server` to support PHP >= 7.1.
2) I'd need `phpactor/language-server` to use ReactPHP stack instead of amphp.

**Pull request closed without merging**